src/fork_gc.c

```diff
@@ -747,58 +811,78 @@ static FGCError recvNumIdx(ForkGC *gc, NumGcInfo *ninfo) {
     goto error;
   }
 
-  // read reduced cardinality size
-  if (FGC_recvFixed(gc, &ninfo->reduceCardinalitySize, sizeof ninfo->reduceCardinalitySize) !=
-      REDISMODULE_OK) {
+  if (recvCardvals(gc, &ninfo->restBlockDeleted, &ninfo->nrestBlockDel) != REDISMODULE_OK) {
     goto error;
   }
-
-  if (ninfo->reduceCardinalitySize) {
-    size_t msize = sizeof(*ninfo->valuesDeleted) * ninfo->reduceCardinalitySize;
-    ninfo->valuesDeleted = rm_malloc(msize);
-    if (FGC_recvFixed(gc, ninfo->valuesDeleted, msize) != REDISMODULE_OK) {
-      goto error;
-    }
+  if (recvCardvals(gc, &ninfo->lastBlockDeleted, &ninfo->nlastBlockDel) != REDISMODULE_OK) {
+    goto error;
   }
 
   return FGC_COLLECTED;
 
 error:
   printf("Error receiving numeric index!\n");
   freeInvIdx(&ninfo->idxbufs, &ninfo->info);
-  rm_free(ninfo->valuesDeleted);
+  rm_free(ninfo->lastBlockDeleted);
+  rm_free(ninfo->restBlockDeleted);
   memset(ninfo, 0, sizeof(*ninfo));
   return FGC_CHILD_ERROR;
 }
 
+static void resetCardinality(NumGcInfo *info, NumericRange *r) {
+  khash_t(cardvals) *kh = kh_init(cardvals);
+  int added;
+  for (size_t ii = 0; ii < info->nrestBlockDel; ++ii) {
+    numUnion u = {info->restBlockDeleted[ii].value};
+    khiter_t it = kh_put(cardvals, kh, u.u64, &added);
+    kh_val(kh, it) = info->restBlockDeleted[ii].appearances;
+  }
+  if (!info->idxbufs.lastBlockIgnored) {
+    for (size_t ii = 0; ii < info->nlastBlockDel; ++ii) {
+      numUnion u = {info->lastBlockDeleted[ii].value};
+      khiter_t it = kh_put(cardvals, kh, u.u64, &added);
+      if (!added) {
+        kh_val(kh, it) += info->lastBlockDeleted[ii].appearances;
+      } else {
+        kh_val(kh, it) = info->lastBlockDeleted[ii].appearances;
+      }
+    }
+  }
+
+  size_t n = array_len(r->values);
+  double minVal = -DBL_MAX, maxVal = DBL_MIN, uniqueSum = 0;
+
+  for (size_t ii = 0; ii < array_len(r->values); ++ii) {
+  reeval:;
+    numUnion u = {r->values[ii].value};
+    khiter_t it = kh_get(cardvals, kh, u.u64);
+    if (it != kh_end(kh) && (r->values[ii].appearances -= kh_val(kh, it)) == 0) {
+      // delet this
+      size_t isLast = array_len(r->values) == ii + 1;
+      array_del_fast(r->values, ii);
+      if (!isLast) {
```

Comment:
Why do you need this? it will continue the loop anyway no? What am I missing?

Reply:
Because a simple continue will increase the iterator value; but after `del_fast` the current iterator will hold a different element, so if we just increase the current iterator, it will skip over.